src/commands/README.md

```diff
@@ -1,13 +1,216 @@
+Command JSON files
+==================
+
 This directory contains JSON files, one for each command.
 
-Each JSON contains all the information about the command itself, but these JSON files are not to be used directly!
-Any third party who needs access to command information must get it from `COMMAND INFO` and `COMMAND DOCS`.
-The output can be extracted in a JSON format by using `redis-cli --json`, in the same manner as in `utils/generate-commands-json.py`.
+Each JSON file contains all the information about the command itself. It is the
+"single source of truth" (SSOT) for the command's metadata.
+
+These JSON files were originally not intended to be used directly, since they
+contain internals and some fields like "acl_categories" that are not the final
+ACL categories. (Valkey will apply some implicit rules to compute the final ACL
+categories.) However, people see JSON files and use them directly anyway.
+
+Any third party who needs access to command information were originally supposed
+to get it from `COMMAND INFO` and `COMMAND DOCS`. These commands can be combined
+into a JSON file by the script `utils/generate-commands-json.py`. Confusingly
+enough, this JSON file as a slightly different format!
+
+Structure
+---------
+
+Each JSON file contains an object with a single key. The key is the command name
+in uppercase, e.g. "HSCAN" (hscan.json). The value is a JSON object with the
+following keys. To be safe, assume all of them are optional.
+
+* `"summary"`: a string with a short description of the command. One sentence.
+* `"complexity"`: a string like `"O(1)"` or longer, like `"O(1) for every call. O(N) for a complete iteration, including enough command calls for the cursor to return back to 0. N is the number of elements inside the collection."`.
+* `"group"`: a string used for categorization in documentation. One of these:
+  * `"bitmap"`
+  * `"cluster"`
+  * `"connection"`
+  * `"generic"`
+  * `"geo"`
+  * `"hash"`
+  * `"hyperloglog"`
+  * `"list"`
+  * `"pubsub"`
+  * `"scripting"`
+  * `"sentinel"`
+  * `"server"`
+  * `"set"`
+  * `"sorted_set"`
+  * `"stream"`
+  * `"string"`
+  * `"transactions"`
+* `"since"`: a string with a version number, like "7.0.0". It's the version
+  (Redis OSS or Valkey) where the command was introduced.
+* `"arity"`: The number of arguments, including the command name itself. A
+  negative number means "at least", e.g. -3 means at least 3.
+* `"container"`: Only present for subcommands. See below.
+* `"history"`: An array of changes, each change represented by a 2-element array
```

Comment:
Might be worth noting that should be omitted if empty. Previously Redis inconsistently used a bunch of `[]` entriee.

Reply:
Sure. That applies to all arrays here, probably. We don't have any validation for it though. We can add that in one of the scripts too.